_posts/2020-06-01-rethinking-the-frontend.md

```diff
@@ -0,0 +1,223 @@
+
+SPA frontend apps are often large applications that the client needs to download before
+it can render the page for the user to see and interact with. This problem has some work
+arounds like module splitting and server side rendering with hydration. These solutions
+bring extra complexity that your application and build process needs to deal with.
+
+With ripley we just send the HTML of the page ready to go. The usual markup for a page in
+the application is likely tiny compared to the frontend JS code and API payloads.
+
+### No need for an API
+
+With SPA frontend you still need to backend to service the data needs.
+Most applications will need to fetch things from a database and store new things to the database.
+This essentially turns the backend into a database API. That's fine if you need it for other clients
+as well, but it shouldn't be necessary just to service your frontend.
+
+The API will also need to encode information to some format (often JSON) and your frontend
+will need to parse the information. With server side rendering this extra round trip can be
+completely avoided.
+
+### No need for client side state management
+
+With the backend being a database API comes another important source of complexity:
+client side state management. You need a way to store the information on the client
+and make sure that long lived pages aren't stale. Programming a two way synchronization
+of frontend app state and the backend database is interesting work to be sure, but the
+complexity is incidental and has nothing to do with the particulars of your application.
+If you can skip it, that's more code that doesn't need to be written and debugged.
+
+### Leveraging browser strengths
+
+SPA frontend apps often require special handling to mimic browser's native navigation
+with [pushState](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState).
+Refreshing in a SPA frontend means losing all the accumulated state and having to
+refetch everything. That's mitigated with frontend routing libraries, but those too
+are another source of incidental complexity.
+
+The browser is very capable of caching your static assets so a server side rendered
+app doesn't need to care about working around
+
+### Build complexity
+
+The traditional server side rendering can be done with a single server project.
+A separate frontend and backend requires two builds each with their own dependencies
+and build steps. Now many would say separation of concerns and differing skill sets
+needed for frontend and backend work makes this split beneficial. That split then
+requires synchronization between the projects or some flexible API like GraphQL.
+
+In my exprience full stack development is very common and the same people are
+doing everything from frontend to backend so the separation argument isn't very
+convincing to me in many cases.
+
+### Server resources
+
+With server side rendering and WebSocket enhancement, the browser is effectively treated
```

Comment:
Perhaps "With WebSocket-enhanced server-side rendering, the…"

Or maybe even: "With WebSocket-enhanced server-side rendering, Ripley effectively treats the browser as a…"